### ISSUE TITLE
bug: mutate + filter compiles to unintentionnal cross join

### ISSUE THREAD

**Issue by BenjaminHabert** [ibis-project/ibis#2744]
The following expression is being compiled to an SQL cross-join although my intention was only to filter the original table.
I am missusing the API or is this a bug ?

```
ibis-framework      1.3.0
Python 3.6.7
```

## Current behavior

```python
person = ibis.table([('person_id', 'int64'), ('birth_datetime', 'timestamp')], name='person')

mutated = (
    person
    .mutate((2021 - person.birth_datetime.year()).name('age'))
)
selected_person = (
    mutated
    .filter(mutated.age <= 40)
    [mutated.person_id, mutated.age]
)

print(ibis.postgres.compile(selected_person))
```
Compiles to the following query which is a cross join (`SELECT FROM [table a], [table b]`).

```sql
SELECT person_id,
       t0.age
FROM
  (SELECT t1.person_id AS person_id,
          t1.birth_datetime AS birth_datetime
   FROM person AS t1),
  (SELECT t1.person_id AS person_id,
          t1.birth_datetime AS birth_datetime,
          :param_1 - CAST(EXTRACT(YEAR
                                  FROM t1.birth_datetime) AS SMALLINT) AS age
   FROM person AS t1) AS t0
WHERE t0.age <= :param_2
```

## Expected behavior

My intention was to filter the original table, corresponding to the following query:

```sql
SELECT t0.person_id,
       t0.age
FROM
  (SELECT t1.person_id AS person_id,
          t1.birth_datetime AS birth_datetime,
          :param_1 - CAST(EXTRACT(YEAR
                                  FROM t1.birth_datetime) AS SMALLINT) AS age
   FROM person AS t1) AS t0
WHERE t0.age <= :param_2
```

Note that small modifications to the python code will indeed compile to the correct query:

```python
person = ibis.table([('person_id', 'int64'), ('birth_datetime', 'timestamp')], name='person')

mutated = (
    person
    # [person.person_id, person.birth_datetime]  # "pre selecting" the columns produces the correct query
    .mutate((2021 - person.birth_datetime.year()).name('age'))
)
selected_person = (
    mutated
    .filter(mutated.age <= 40)
    [mutated.person_id, mutated.age]  #  using ["person_id", "age"]  instead of the column expressions
                                                              # will also compile to the correct query
)

print(ibis.postgres.compile(selected_person))
```

**Comment by cpcloud** [comment]
This is definitely a bug! Thanks for posting.

**Comment by cpcloud** [comment]
@BenjaminHabert I looked into this yesterday and it's a fairly tricky bug to fix. However, we're still aiming to fix it for 3.0.0.

In the meantime, you can workaround the issue by referring to the columns in the final projection in your example using strings:

```python
person = ibis.table([('person_id', 'int64'), ('birth_datetime', 'timestamp')], name='person')

mutated = (
    person
    .mutate((2021 - person.birth_datetime.year()).name('age'))
)
selected_person = (
    mutated
    .filter(mutated.age <= 40)
    ["person_id", "age"]
)

print(ibis.postgres.compile(selected_person))
```
